programs/cleos/main.cpp

```diff
@@ -519,9 +519,11 @@ fc::variant push_transaction( signed_transaction& trx, const std::vector<public_
          }
       } else {
          EOSC_ASSERT( !tx_unpack_data, "ERROR: --unpack-action-data not supported with --return-packed" );
-        return fc::variant(packed_transaction(trx, compression));
+         return fc::variant(packed_transaction(trx, compression));
       }
    }
+   assert(0);
+   return {}; // fix warning: control reaches end of non-void function [-Wreturn-type]
```

Comment:
What compiler is complaining here? It doesn't seem any of the builds in CI nor a local gcc12 complains, and it seems like `Wreturn-type` is enabled. Is this the best way to silence a warning such as this? Is `__builtin_unreachable()` any better or worse?

Reply:
My compiler (on ubuntu 22.04.1) is gcc 11.3.0.
I don't think `__builtin_unreachable()`  is a good alternative. First it is an optimization pragma, not an assertion. And if some future code change allows the code to reach here, we should have a valid return statement rather that returning some random value from the stack.

Reply:
oh, I see, you don't get the warning until adding the `assert(0)`

Reply:
No, I did get the warning, which was why I assed the `assert()` and the `return`

Reply:
When removing both lines (the assert and the return):

```
[2/3] /usr/bin/c++ -DBOOST_ALL_NO_LIB -DEOSIO_EOS_VM_JIT_RUNTIME_ENABLED -DEOSIO_EOS_VM_OC_RUNTIME_ENABLED -DEOSIO_EOS_VM_RUNTIME_ENABLED -DEOS_VM_SOFTFLOAT -DOPENSSL_API_COMPAT=0x10100000L -DOPENSSL_NO_DEPRECATED -I/home/greg/github/enf/leap/build_debug/programs/cleos -I/home/greg/github/enf/leap/programs/cleos -I/home/greg/github/enf/leap/libraries/appbase/include -I/home/greg/github/enf/leap/libraries/version/include -I/home/greg/github/enf/leap/libraries/cli11/include -I/home/greg/github/enf/leap/plugins/chain_api_plugin/include -I/home/greg/github/enf/leap/plugins/chain_plugin/include -I/home/greg/github/enf/leap/plugins/chain_plugin/../chain_interface/include -I/home/greg/github/enf/leap/plugins/chain_plugin/../../libraries/appbase/include -I/home/greg/github/enf/leap/plugins/chain_plugin/../producer_plugin/include -I/home/greg/github/enf/leap/plugins/chain_plugin/../resource_monitor_plugin/include -I/home/greg/github/enf/leap/plugins/producer_plugin/include -I/home/greg/github/enf/leap/plugins/producer_plugin/../chain_interface/include -I/home/greg/github/enf/leap/plugins/http_client_plugin/include -I/home/greg/github/enf/leap/libraries/chain/include -I/home/greg/github/enf/leap/build_debug/libraries/chain/include -I/home/greg/github/enf/leap/libraries/chain/../wasm-jit/Include -I/home/greg/github/enf/leap/libraries/libfc/include -I/home/greg/github/enf/leap/libraries/libfc/libraries/ff/libff/.. -I/home/greg/github/enf/leap/libraries/libfc/secp256k1/secp256k1 -I/home/greg/github/enf/leap/libraries/libfc/secp256k1/secp256k1/include -I/home/greg/github/enf/leap/libraries/chainbase/include -I/home/greg/github/enf/leap/libraries/chain/../rapidjson/include -I/home/greg/github/enf/leap/libraries/wasm-jit/Source/Runtime/../../../chain/include -I/home/greg/github/enf/leap/libraries/softfloat/source/include -I/home/greg/github/enf/leap/libraries/softfloat/source/8086-SSE -I/home/greg/github/enf/leap/libraries/softfloat/build/Linux-x86_64-GCC -I/home/greg/github/enf/leap/libraries/builtins -I/home/greg/github/enf/leap/libraries/builtins../softfloat/source/include -I/home/greg/github/enf/leap/libraries/eos-vm/include -I/home/greg/github/enf/leap/libraries/eos-vm/external/softfloat/source/include -I/home/greg/github/enf/leap/plugins/signature_provider_plugin/include -I/home/greg/github/enf/leap/plugins/resource_monitor_plugin/include -I/home/greg/github/enf/leap/plugins/http_plugin/include -Wall -g -fdiagnostics-color=always -Wno-register -std=gnu++17 -MD -MT programs/cleos/CMakeFiles/cleos.dir/main.cpp.o -MF programs/cleos/CMakeFiles/cleos.dir/main.cpp.o.d -o programs/cleos/CMakeFiles/cleos.dir/main.cpp.o -c /home/greg/github/enf/leap/programs/cleos/main.cpp
/home/greg/github/enf/leap/programs/cleos/main.cpp: In function ‘fc::variant push_transaction(eosio::chain::signed_transaction&, const std::vector<fc::crypto::public_key>&)’:
/home/greg/github/enf/leap/programs/cleos/main.cpp:525:1: warning: control reaches end of non-void function [-Wreturn-type]
  525 | }
      | ^

❯ /usr/bin/c++ --version                                                                                                   20s 
c++ (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Reply:
The latest ubuntu22 build on master doesn't seem to have a warning there, so not sure what is causing the difference, it's possibly Debug vs Release related
https://github.com/AntelopeIO/leap/actions/runs/3664388649/jobs/6194769231

Reply:
Oh yes, that is probably the reason, I am building in debug mode.
But regardless of  why the warning is obtained, I don't think it hurts to address it.

Reply:
`assert` is only enabled in debug. If control ever reaches here, it is a bug. Why not just print out the error and exit?

Reply:
In general we use `FC_ASSERT` or `EOS_ASSERT`

Reply:
I can make the change to `FC_ASSERT` or `EOS_ASSERT`, but I am a bit surprised that these throw an exception.

I believe throwing exception is appropriate for conditions that are expected to happen in exceptional conditions (for example memory allocation failing). In these cases we throw an exception to 'abort' the current action, with the hope that the exception will be caught and that the stack unrolling will free some of the resource crunch causing the exception. It may be viewed as a normal program occurence.

However sometimes there are conditions that logically should never happen (loop invariants for example, preconditions, etc...), and you feel so strongly about that that you assert that condition in the code. You effectively state, this will **always** be true, and if you are wrong you want to be immediately notified in the loudest and most immediate way. 
In those cases you do *not* want to throw an exception, that can be gracefully caught at a higher level and silenced.

Anyways, that is how I understand asserts(). Just thought I'd mention it.

Reply:
`assert()` is a noop in release builds (`NDEBUG`), that is pretty sneaky and not desirable for many places we'd use `FC_ASSERT` as a "this should never happen"

Reply:
imho `FC_ASSERT` has a better chance of being ignored as the throw can easily be caught and be invisible. That is why for a condition that "should never happen" a hard termination is better. 
And because it should never happen, it should have no cost in release mode, not even a test and code branch.

Reply:
We currently do not build/run debug mode in CI/CD.

Reply:
But developers do run the code under the debugger, don't they? And they run the unit tests with the debug build as well (I do). That should be enough to trigger the asserts if they fail I think.
If not running CI/CD on a debug build is the only objection, then we definitely should indeed change this. The benefit is that we free the release build from testing conditions that will never occur, and empower the developer to add as many asserts he/she wants, knowing there is no cost in release builds.

Reply:
@spoonincode @heifner do you still think I should change the `assert()` to `FC_ASSERT()`?

Reply:
The
```c++
   assert(0);
   return {};
```
isn't an established pattern in Leap, so it doesn't seem appropriate here imo. `FC_ASSERT()` is more in line with common patterns. (I also generally don't like adding cruft to silence spurious warnings such as this, however I can sympathize with the angle that this is a long function that's hard to reason through if all branches return so it ought to have the clear assertion)

Reply:
I vote with FC_ASSERT.

Reply:
Thank you @spoonincode and @linh2931 . I had a long discussion yesterday evening with Areg and we arrived at a different conclusion. Areg suggested we talk about this in the engineering meeting on Tuesday, and I have added this topic to the agenda.
So for now, I might just remove this change (it was only to address a compiler warning) until we come to an agreement.

Reply:
After talking with Areg, I removed the change, there may be a better resolution for this warning by rearranging the code.